src/rust/generator-executor/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+use async_stream::stream;
+use futures_util::pin_mut;
+use generator_sdk::{
+    client::{
+        CacheBuilder,
+        GeneratorClient,
+        GeneratorClientError,
+        TokioAsyncResolver,
+    },
+    ClientConfig as GeneratorClientConfig,
+};
+use grapl_utils::future_ext::GraplFutureExt;
+use plugin_work_queue::client::PluginWorkQueueServiceClient;
+use rust_proto::plugin_work_queue::{
+    AcknowledgeGeneratorRequest,
+    ExecutionJob,
+    GetExecuteGeneratorRequest,
+    GetExecuteGeneratorResponse,
+};
+use structopt::StructOpt;
+use tokio::time::{
+    sleep,
+    Duration,
+};
+use tokio_stream::{
+    Stream,
+    StreamExt,
+};
+
+#[derive(StructOpt)]
+pub struct GeneratorExecutorConfig {
+    #[structopt(env)]
+    plugin_work_queue_address: String,
+    // Path to the certificate to be used when validating the generator plugin TLS sessions
+    #[structopt(env)]
+    generator_certificate_path: std::path::PathBuf,
+    #[structopt(flatten)]
+    generator_client_config: GeneratorClientConfig,
+}
+
+fn make_job_stream(
+    mut work_queue_client: PluginWorkQueueServiceClient<tonic::transport::Channel>,
+) -> impl Stream<Item = (ExecutionJob, i64)> {
+    stream! {
+        loop {
+            let message = match work_queue_client.get_execute_generator(GetExecuteGeneratorRequest {}).await {
+                Ok(message) => message,
+                Err(e) => {
+                    tracing::error!(message="Failed to retrieve generator job", error=?e);
+                    // check what kind of error it is and retry it accordingly
+                    sleep(Duration::from_secs(1)).await;
+                    continue
+                }
+            };
+            match message {
+                GetExecuteGeneratorResponse { execution_job: Some(job), request_id } => yield (job, request_id),
+                _ => {
+                    tracing::debug!(message="No available jobs");
+                    sleep(Duration::from_secs(1)).await;
+                }
+            };
+
+        }
+    }
+}
+
+// Attempts to ack the message. On failure, will retry in a background task
+async fn try_ack(
+    work_queue_client: &mut PluginWorkQueueServiceClient<tonic::transport::Channel>,
+    request_id: i64,
+    success: bool,
```

Comment:
Something I've often found useful from a readability / maintenance standpoint is to replace bare `bool` values with descriptive enums.

`try_ack(client, id, Status::Success)` and `try_ack(client, id, Status::Failure)` read much more easily than the identical code with simply `true` and `false`.

Alternatively (and probably better, in this case, honestly), you could have `ack_success` and `ack_failure` as thin facades over `try_ack` that handle the booleans internally, leaving the broader code easier to read.

(Idiomatically, it also seems odd to call this function `try_XXX` when it doesn't return a `Result`.)